awsx/ec2/vpc.ts

```diff
@@ -0,0 +1,273 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+import * as schema from "../schema-types";
+import { getSubnetSpecs, SubnetSpec } from "./subnetDistributor";
+
+interface VpcData {
+    vpc: aws.ec2.Vpc;
+    subnets: aws.ec2.Subnet[];
+    routeTables: aws.ec2.RouteTable[];
+    routes: aws.ec2.Route[];
+    routeTableAssociations: aws.ec2.RouteTableAssociation[];
+    igw: aws.ec2.InternetGateway;
+    natGateways: aws.ec2.NatGateway[];
+    eips: aws.ec2.Eip[];
+}
+
+export class Vpc extends schema.Vpc<VpcData> {
+    constructor(name: string, args: schema.VpcArgs, opts: pulumi.ComponentResourceOptions = {}) {
+        super(name, args, opts);
+
+        const data = pulumi.output(this.getData());
+
+        this.vpc = data.vpc;
+        this.subnets = data.subnets;
+        this.routeTables = data.routeTables;
+        this.routes = data.routes;
+        this.routeTableAssociations = data.routeTableAssociations;
+        this.internetGateway = data.igw;
+        this.natGateways = data.natGateways;
+        this.eips = data.eips;
+    }
+
+    protected async initialize(props: { name: string, args: schema.VpcArgs, opts: pulumi.ComponentResourceOptions; }) {
+        const { name, args } = props;
+        const availabilityZones = args.availabilityZoneNames ?? await this.getDefaultAzs();
+
+        const natGatewayStrategy: schema.NatGatewayStrategyInputs = args.natGateways?.strategy ?? "OnePerAz";
+        const allocationIds = args.natGateways?.elasticIpAllocationIds ?? [];
+        validateEips(natGatewayStrategy, allocationIds, availabilityZones);
+
+        const cidrBlock = args.cidrBlock ?? "10.0.0.0/16";
+
+        const subnetSpecs = getSubnetSpecs(name, cidrBlock, availabilityZones, args.subnetsPerAz);
+        validateNatGatewayStrategy(natGatewayStrategy, subnetSpecs);
+
+        const vpc = new aws.ec2.Vpc(
+            name,
+            {
+                ...args,
+                cidrBlock,
+                tags: {
+                    ...args.tags,
+                    Name: name,
+                },
+            },
+            { parent: this },
+        );
+
+        // We unconditionally create the IGW (even if it's not needed because we
+        // only have isolated subnets) because AWS does not charge for it, and
+        // therefore there's no harm in adding it, whereas conditional resources
+        // must be declared in the constructor, which significantly complicates the
+        // code.
+        const igw = new aws.ec2.InternetGateway(`${name}`, {
+            vpcId: vpc.id,
+        }, { parent: this });
```

Comment:
the parent of this should be `this.vpc.id` I think

Reply:
Agreed.

Reply:
Will be in next push.